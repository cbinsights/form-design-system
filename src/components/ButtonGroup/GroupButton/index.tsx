import React from 'react';
import cc from 'classcat';
import baseElement from 'util/baseElement';
import Icon from 'components/Icon';

const WrapperComponent = ({ children }: { children: JSX.Element }): JSX.Element =>
  children;

export interface GroupButtonProps {
  /** Displays active button style when true */
  isActive?: boolean;

  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: JSX.Element;

  /** Text that appears in GroupButton */
  label?: string;

  /** Controls whether button is disabled or not (works for
   * anchor tags as well, visually) */
  disabled?: boolean;

  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon?: Icon;

  /**
   * A component to wrap the entire group button. The
   * `wrapper` must render the children prop passed into it
   * to render the GroupButton.
   */
  Wrapper?: (props: { children: React.ReactNode }) => JSX.Element;

  /** Assigns the first button styling. */
  isFirstButton?: boolean;

  /** Assigns the last button styling. */
  isLastButton?: boolean;

  /** Specifies a URL to link to.  */
  href?: string;

  onClick: () => void;
}

export const GroupButton = (props: GroupButtonProps): JSX.Element => {
  const {
    Wrapper = WrapperComponent,
    isActive,
    Icon,
    label,
    Link,
    isFirstButton,
    isLastButton,
    href,
    ...rest
  } = props;

  const Element = baseElement({ href: href, onClick: true, as: Link });

  const rootClass = cc([
    {
      'groupbtn--disabled': props.disabled,
      'groupbtn--active': isActive,
      'groupbtn--first': isFirstButton,
      'groupbtn--last': isLastButton,
    },
    'groupbtn',
    'border--focus--noTransition',
    'transition--default',
  ]);

  return (
    <Wrapper>
      <Element {...rest} className={rootClass}>
        {Icon && (
          <div className="margin--right--s">
            <Icon size="xs" />
          </div>
        )}
        <span className="groupbtn-label">{label}</span>
      </Element>
    </Wrapper>
  );
};

export default GroupButton;
