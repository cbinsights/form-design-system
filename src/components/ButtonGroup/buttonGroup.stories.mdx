import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, object } from '@storybook/addon-knobs';
import Updatable from 'components/Updatable'
import Tooltip from 'components/Tooltip'
import FeedIcon from 'lib/icons/react/FeedIcon';

import ButtonGroup from '.';
import GroupButton from './GroupButton';

import { ImportPath, State } from 'components/util/storybook';

export const buttons = [
  { label: 'Lorem', Icon: FeedIcon, isActive: true },
  { label: 'Ipsum', Icon: FeedIcon, disabled: true },
  { label: 'Dolor', Icon: FeedIcon },
  { label: 'Sit', Icon: FeedIcon },
];

export const buttonsNoIcons = [
  { label: 'Lorem', disabled: true },
  { label: 'Ipsum' },
  { label: 'Dolor', isActive: true },
  { label: 'Sit' },
];

<Meta title="Components/ButtonGroup" decorators={[withKnobs]} component={ButtonGroup} />

# ButtonGroup

<ImportPath component={ButtonGroup} section="interactive" />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <ButtonGroup buttons={object('buttons', buttonsNoIcons)} />
</Story>

<Preview isExpanded>
  <ButtonGroup
    buttons={[
      { label: 'Lorem', disabled: true },
      { label: 'Ipsum' },
      { label: 'Dolor', isActive: true }
    ]} 
  />
</Preview>

<Props of={ButtonGroup} />
<Props of={GroupButton} />

### Button States

<Preview>
  <Story name="States">
    <State
      initialValue={1}
      onChange={console.log}
      render={((value, setValue) => {
        return (
          <ButtonGroup
            buttons={
              ['Lorem', 'Sit', 'Ipsum'].map((label, idx) => ({
                label: label, onClick: () => setValue(idx), isActive: value === idx
              }))
            }
          />
        )
      })}
    />
  </Story>
</Preview>

### Wrapper
Allows you to render a component around any of the individual buttons

<Preview>
  <Story name="Wrapper">
    <ButtonGroup buttons={[
      { label: 'Updatable', Wrapper: (props) => (<Updatable message="42">{props.children}</Updatable>) },
      { label: 'Tooltip',  Wrapper: (props) => (
        <Tooltip trigger={props.children} message="hey" maxWidth={240} />
      )},
    ]} 
    />
  </Story>
</Preview>



### Icons
(Displays on the left)

<Preview>
  <Story name="Icons">
    <ButtonGroup buttons={[
      { label: 'Lorem', Icon: FeedIcon },
      { label: 'Sit', Icon: FeedIcon  },
    ]} />
  </Story>
</Preview>
