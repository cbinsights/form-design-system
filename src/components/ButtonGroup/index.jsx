import React from 'react';
import PropTypes from 'prop-types';
import combine from 'util/combine';
import GroupButton from './GroupButton';

const ButtonGroup = ({ buttons, onChange, ...restBtnGroup }) => (
  <div {...restBtnGroup} className="btngroup">
    {buttons.map(
      (
        {
          value,
          content,
          key,
          Icon,
          onClick,
          label,
          // `isFirstButton` and `isLastButton` are private and can't being passed in.
          isFirstButton,
          isLastButton,
          ...restBtn
        },
        index
      ) => (
        <GroupButton
          key={key || label}
          onClick={onChange ? combine(() => onChange(label || value), onClick) : onClick}
          label={label}
          Icon={Icon}
          isFirstButton={index === 0}
          isLastButton={index === buttons.length - 1}
          {...restBtn}
        />
      )
    )}
  </div>
);

ButtonGroup.propTypes = {
  /**
   * Each object in array renders a GroupButton, and (generally) passes all it's values through
   * as props to GroupButton
   */
  buttons: PropTypes.arrayOf(PropTypes.shape(GroupButton.propTypes)).isRequired,

  /**
   * Fires when any Button is clicked in ButtonGroup
   */
  onChange: PropTypes.func,
};

export default ButtonGroup;
