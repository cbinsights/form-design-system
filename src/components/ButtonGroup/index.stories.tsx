import React from 'react';
import Tooltip from 'components/Tooltip';
import FeedIcon from 'icons/react/FeedIcon';
import State from 'util/storybook-docs/State';
import { StoryObj } from '@storybook/react';
import ButtonGroup, { ButtonGroupProps } from '.';
import { noop } from 'util/index';
export const Primary: StoryObj<ButtonGroupProps> = {
  args: {
    buttons: [
      {
        label: 'Lorem',
        value: 1,
        key: 1,
        disabled: true,
        onClick: noop,
      },
      {
        label: 'Ipsum',
        value: 2,
        key: 2,
        onClick: noop,
      },
      {
        label: 'Dolor',
        value: 3,
        key: 3,
        isActive: true,
        onClick: noop,
      },
    ],
  },
};

export const ActiveStates = (): JSX.Element => (
  <State
    initialValue={1}
    render={(value, setValue) => (
      <ButtonGroup
        buttons={['Lorem', 'Sit', 'Ipsum'].map((label, idx) => ({
          label,
          onClick: () => setValue(idx),
          isActive: value === idx,
          key: idx,
          value: idx,
        }))}
      />
    )}
  />
);

export const Wrapper: StoryObj<ButtonGroupProps> = {
  args: {
    buttons: [
      {
        label: 'Tooltip',
        onClick: noop,
        key: 1,
        value: 1,
        Wrapper: (props) => (
          <Tooltip trigger={<>{props.children}</>} message="hey" maxWidth={240} />
        ),
      },
    ],
  },
  parameters: {
    docs: {
      description: {
        story: 'Allows you to render a component around any of the individual buttons',
      },
    },
  },
};

export const Icons: StoryObj<ButtonGroupProps> = {
  args: {
    buttons: [
      {
        label: 'Lorem',
        Icon: FeedIcon,
        onClick: noop,
        value: 1,
        key: 1,
      },
      {
        label: 'Sit',
        Icon: FeedIcon,
        onClick: noop,
        value: 2,
        key: 2,
      },
    ],
  },
  parameters: {
    docs: {
      description: {
        story: 'Displays on the left',
      },
    },
  },
};

export default {
  component: ButtonGroup,
  title: 'Components/ButtonGroup',
};
