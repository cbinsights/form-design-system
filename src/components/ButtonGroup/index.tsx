import React from 'react';
import GroupButton, { GroupButtonProps } from './GroupButton';

type ButtonModificationProps = {
  key: string | number;
  value: number;
  onClick: any;
};

type CombinedButtonProps = GroupButtonProps & ButtonModificationProps;
export interface ButtonGroupProps {
  /**
   * Each object in array renders a GroupButton, and (generally) passes all it's values through
   * as props to GroupButton
   */
  buttons: CombinedButtonProps[];

  /**
   * Fires when any Button is clicked in ButtonGroup
   */
  onChange?: any;
}

const ButtonGroup = ({
  buttons,
  onChange,
  ...restBtnGroup
}: ButtonGroupProps): JSX.Element => (
  <div {...restBtnGroup} className="btngroup">
    {buttons.map(({ value, key, Icon, onClick, label, ...restBtn }, index) => (
      <GroupButton
        key={key || label}
        onClick={onChange}
        label={label}
        Icon={Icon}
        isFirstButton={index === 0}
        isLastButton={index === buttons.length - 1}
        {...restBtn}
      />
    ))}
  </div>
);

export default ButtonGroup;
