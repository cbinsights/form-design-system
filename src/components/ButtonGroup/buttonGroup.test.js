import React from 'react';
import { shallow } from 'enzyme';

import ButtonGroup from '.';

const renderComponent = (props) => shallow(<ButtonGroup {...props} />);

const Icon = () => (
  <span>ðŸŒ­</span>
)

const buttons = [
  { label: 'Feed', Icon, isActive: true },
  { label: 'Table', Icon, onClick: () => {}, disabled: true },
];

describe('ButtonGroup component', () => {

  it('matches snapshot (default props)', () => {
    const component = renderComponent({
      buttons,
    });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent({
      buttons,
      onChange: () => {}
    });
    expect(component).toMatchSnapshot();
  });
});
