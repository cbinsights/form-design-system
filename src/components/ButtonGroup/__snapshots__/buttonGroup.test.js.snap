// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup component matches snapshot (default props) 1`] = `
<div
  className="btngroup"
>
  <GroupButton
    Icon={[Function]}
    isActive={true}
    isFirstButton={true}
    isLastButton={false}
    key="Feed"
    label="Feed"
  />
  <GroupButton
    Icon={[Function]}
    disabled={true}
    isFirstButton={false}
    isLastButton={true}
    key="Table"
    label="Table"
    onClick={[Function]}
  />
</div>
`;

exports[`ButtonGroup component matches snapshot (set all props) 1`] = `
<div
  className="btngroup"
>
  <GroupButton
    Icon={[Function]}
    isActive={true}
    isFirstButton={true}
    isLastButton={false}
    key="Feed"
    label="Feed"
    onClick={[Function]}
  />
  <GroupButton
    Icon={[Function]}
    disabled={true}
    isFirstButton={false}
    isLastButton={true}
    key="Table"
    label="Table"
    onClick={[Function]}
  />
</div>
`;
